### ISSUE TITLE
Model fit in Parallel.For is slower than For

### ISSUE THREAD

**Issue by huy302** [dotnet/machinelearning#4524]
### System information

- **OS version/distro**: Win10 64
- **.NET Version (eg., dotnet --info)**: 4.6.1

### Issue

- **What did you do?**
I'm trying to tune hyperparamater and see if I can implement grid search myself. In order to do that, I would like to fit the model in multiple threads by using Parallel.For

- **What happened?**
However, I realized that fitting the model multiple times using Parallel.For takes much longer than a single threaded for loop. In a for loop. my CPU utilization is at around 30% all the time, while using Parallel.For, it spikes to 100% at first, then drops down to 6% for quite some time, then jumps back to 30% and finishes. Am I not supposed to fit in parallel? If that is the case then what is the best way to do grid search?

### Source code / logs

Single threaded for - takes 20 seconds

```
for (var i = 0; i < 100; i++) {
    var pipeline = CreatePipeline();
    pipeline.Fit(dataView);
}
```

Parallel.For - takes 56 seconds

```
Parallel.For(0, 100, i => {
    var pipeline = CreatePipeline();
    pipeline.Fit(dataView);
});
```

**Comment by gvashishtha** [comment]
@frank-dong-ms

**Comment by frank-dong-ms** [comment]
Thanks for reaching out, can you share sample code that I can reproduce the issue in my local and investigate more?

**Comment by huy302** [comment]
You can reproduce it with any provided sample codes. For example I'm using the code from getting-started\MulticlassClassification_MNIST\Program.cs with a small modification in Main()

`
        static void Main(string[] args)
        {
            var sw = new Stopwatch();
            sw.Start();
            for (var i = 0; i < 5; i++)
            {
                MLContext mlContext = new MLContext();
                Train(mlContext);
            }
            sw.Stop();
            Console.WriteLine("Sequential: " + sw.Elapsed);

            sw.Restart();
            Parallel.For(0, 5, i =>
            {
                MLContext mlContext = new MLContext();
                Train(mlContext);
            });
            sw.Stop();
            Console.WriteLine("Parallel: " + sw.Elapsed);

            //TestSomePredictions(mlContext);

            Console.WriteLine("Hit any key to finish the app");
            Console.ReadKey();
        }
`

**Output:**

Sequential: 00:00:51.4589381

Parallel: 00:02:56.3580363

**Comment by frank-dong-ms** [comment]
I have a repro in my local and investigating the root cause. For hyper-parameter tuning, you don't need to do it your self, please check out automl: https://www.microsoft.com/en-us/research/project/automl/

**Comment by huy302** [comment]
Hi Frank, I'm aware of the automl feature and will use that for tuning
However, there are some other workflows that we would like to run prediction in parallel as well
Any luck on the performance degradation issue when using parallel for?
Thanks

**Comment by frank-dong-ms** [comment]
The issue with parallel seems like ML.NET is opening too much threads during execution thus impacting the performance of parallel. This can be a quite large issue as the fix will touch some foundation code in ML.NET, we will update if we have any update. Thanks!

**Comment by huy302** [comment]
thanks for the update Frank, cheers

**Comment by frank-dong-ms** [comment]
@huy302 I have done more tests on this issue and below is my test result:

From ML.NET, there are mainly 2 place have multi-threading enabled, one is Text Loader and the other is DataViewUtils, I'm running 20 iteration for sequential and parallel with threading enabled for disabled from Text Loader and DataViewUtils:


1. Threading enabled:
Sequential: 00:02:39.7943361
Parallel:      00:21:09.2656613

2. TextLoader Threading disabled:
Sequential: 00:02:41.2897879
Parallel:      00:20:05.1208179


3. DataViewUtils Threading disabled:
Sequential: 00:07:02.6246745
Parallel:      00:06:16.3059868


4. all Threading disabled:
Sequential: 00:06:56.0748321
Parallel:      00:06:42.4875673


And here is my conclusion from this test:
1. Threading in Text loader not have much impact on overall performance 
2. Threading in DataViewUtils has major performance impact(sequential from 7 minutes to less than 3 minutes). From DataViewUtils we are using threads count equals number of process, so extra threading introduced from Parallel.For not enhance performance but make the execution much slower. I'm not fully understand why Parallel.For is making the execution 10x slower maybe there are many context switch that are expensive.
3. Since ML.NET already has multi-threading during training and inference, we don't need to introduce extra threading by Parallel.For as we are already take advantage of threading.

Please let me know if this explains your question and make sense to you, thanks.

**Comment by frank-dong-ms** [comment]
This is by design, close the issue for now, please feel free to reopen if you have more concerns on this issue, thanks.
